pkg/relayer/session/session.go

```diff
@@ -123,17 +132,22 @@ func (rs *relayerSessionsManager) Start(ctx context.Context) {
 // and/or ensure that the state at each pipeline stage is persisted to disk
 // and exit as early as possible.
 func (rs *relayerSessionsManager) Stop() {
-	rs.relayObs.UnsubscribeAll()
+	for _, v := range rs.relayObs {
+		v.UnsubscribeAll()
+	}
 }
 
 // SessionsToClaim returns an observable that notifies when sessions are ready to be claimed.
 func (rs *relayerSessionsManager) InsertRelays(relays relayer.MinedRelaysObservable) {
+	// TODO_IN_THIS_PR: potentially use a map[keyname/address]relayObs instead
+	// TODO_IN_THIS_COMMIT: figure out how to get an address from relays, or add an argument to `InsertRelays`.
```

Comment:
@red-0ne what do you think the best way to get the address here?

Reply:
Here is my thought procs:

1. `relayerSessionsManager` is a receiver. I wonder what's inside of that?
2. I went to `sesison.go` and found:

```go
type relayerSessionsManager struct {
	logger polylog.Logger

	relayObs relayer.MinedRelaysObservable

	// sessionsToClaimObs notifies about sessions that are ready to be claimed.
	sessionsToClaimObs observable.Observable[[]relayer.SessionTree]

	// sessionTrees is a map of block heights pointing to a map of SessionTrees
	// indexed by their sessionId.
	// The block height index is used to know when the sessions contained in the entry should be closed,
	// this helps to avoid iterating over all sessionsTrees to check if they are ready to be closed.
	sessionsTrees   sessionsTreesMap
	sessionsTreesMu *sync.Mutex

	// blockClient is used to get the notifications of committed blocks.
	blockClient client.BlockClient

	// supplierClient is used to create claims and submit proofs for sessions.
	supplierClient client.SupplierClient

	// pendingTxMu is used to prevent concurrent txs with the same sequence number.
	pendingTxMu sync.Mutex

	// storesDirectory points to a path on disk where KVStore data files are created.
	storesDirectory string

	// sharedQueryClient is used to query shared module parameters.
	sharedQueryClient client.SharedQueryClient
}
```
3. I found a bunch of clients such as `sharedQueryClient client.SharedQueryClient`
4. I started wondering what other clients exist so I clicked into `client.SharedQueryClient`
5. That took me to `interface.go`
6. I did a ctrl-f for `AccountQueryClient` and found this:
```go
// AccountQueryClient defines an interface that enables the querying of the
// on-chain account information
type AccountQueryClient interface {
	// GetAccount queries the chain for the details of the account provided
	GetAccount(ctx context.Context, address string) (cosmostypes.AccountI, error)

	// GetPubKeyFromAddress returns the public key of the given address.
	GetPubKeyFromAddress(ctx context.Context, address string) (cryptotypes.PubKey, error)
}
```
7. My intuition is telling me that I just need to update the `relayerSessionsManager` to inject an instance of `AccountQueryClient`.
8. I would look at how this is done in other part of the codebase and copy-pasta

@okdas Does that help?

Reply:
Not sure what you want to achieve here, but I don't think it will be useful to try to split the relays by `Supplier` there.

You can notice that this pipe has a dead end at `mapAddMinedRelayToSessionTree`. Which is just adding `Relay`s to its corresponding `SMT`.

A better place where you can group `SessionTree`s by `Supplier` would be at `newMapClainSessionsFn`.

There [1] you could have `sessionClaims` to be a `map[supplierAddress][]*relayer.SessionClaim` instead of `[]*relayer.SessionClaim`

And do (error handling is omitted)
```go
for _, session := range sessionTrees {
  address := session.supplierAddress
  sessionClaims[address] = append(sessionClaims[address], relayer.SessionClaim)
}

for supplierAddress, supplierClaims := sessionClaims {
  rs.supplierClientMap[supplierAddress].CreateClaims(ctx, supplierclaims)
}
```

You do the same thing for proofs [2]

[1] https://github.com/pokt-network/poktroll/blob/main/pkg/relayer/session/claim.go#L155
[2] https://github.com/pokt-network/poktroll/blob/main/pkg/relayer/session/proof.go#L151

Reply:
@red-0ne thanks - that's precisely what I planned to do (there's a todo in that section of the file). I originally thought we'd also need to group the observables, but it doesn't seem to be the case. Thank you! I think I'm getting really close now and will send a PR for a review before my EOD.